### PR TITLE
feat(smart-action): users can now display html when an error occurs

### PR DESCRIPTION
**Pull request by SteveBunlon**
### General

- [x] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [x] Test manually the implemented changes
- [x] Validate the code quality (indentation, syntax, style, simplicity, readability)
- [x] Ensure that Types have been updated according to your changes (if needed)

### Security

- [x] Consider the security impact of the changes made

### CODE REVIEW

**Comment by SteveBunlon** [comment]
This PR is not finished yet

File: packages/datasource-toolkit/src/decorators/actions/result-builder.ts [hunk 1]
```diff
@@ -25,13 +25,15 @@ export default class ResultBuilder {
   /**
    * Returns an error response from the action
    * @param message the error message to return
+   * @param html the html content to display
    * @example
    * .error('Failed to refund the customer!');
    */
-  error(message?: string): ActionResult {
+  error(message?: string, html?: string): ActionResult {
```
Comment (romain-gilliotte) [inline packages/datasource-toolkit/src/decorators/actions/result-builder.ts:32]: Please use the same syntax than for success (first param is the message, second param is options).

**Review by romain-gilliotte — CHANGES_REQUESTED** [review]
If the frontend supports the same for errors and success, please reuse the `SuccessResult` type for both errors and success (rename it to `DisplayResult`) and factorize everything

**Comment by romain-gilliotte** [comment]
> This PR is not finished yet

We use github draft on this repo when a PR is not ready :)
I changing the state of this one to draft

File: packages/datasource-toolkit/src/decorators/actions/result-builder.ts (hunk 1 above)
Comment (SteveBunlon) [reply packages/datasource-toolkit/src/decorators/actions/result-builder.ts:32]: By looking at the success code, you can only customise message OR html, not both. That's a problem. Any insight ?

File: docs/agent-customization/actions/responses.md [hunk 2]
```diff
@@ -44,6 +44,21 @@ return resultBuilder.success(
...
+  'You can perform this action for the following reasons:',
+  `
+    <p class="c-clr-1-4 l-mt l-mb">\$${record.amount / 100} USD exceeds threshold.</p>
+    <strong class="c-form__label--read c-clr-1-2">You can use the following card instead</strong>
+    <p class="c-clr-1-4 l-mb">**** **** **** ${record.source.last4}</p>
+  `,
+  { type: 'html' },
+);
+```
+
```
Comment (SteveBunlon) [inline docs/agent-customization/actions/responses.md:61]: Screenshot missing here

File: packages/datasource-toolkit/src/decorators/actions/result-builder.ts [hunk 3]
```diff
@@ -8,30 +8,29 @@ export default class ResultBuilder {
...
    * @param options available options to return
    * @example
-   * .success('<blinkee>Success!</blinkee>', { type: 'html' });
+   * .success('Success', { html: '<blinkee>Success!</blinkee>' });
    */
-  success(
-    message?: string,
-    options?: { type?: 'html' | 'text'; invalidated?: string[] },
-  ): ActionResult {
+  success(message?: string, options?: { html?: string; invalidated?: string[] }): ActionResult {
```
Comment (romain-gilliotte) [inline packages/datasource-toolkit/src/decorators/actions/result-builder.ts:13]: can we rename `message` to `toasterMessage`
and `options.html` to `options.sidePanelHtml`?

Something like this:
```
  /**
   * Returns a success response from the action
   * @param toasterMessage the success message to display on the toaster
   * @param options.sidePanelHtml in addition to the toaster, you may open a side-panel to display
   *   the result in HTML
   * @param options.invalidated for summary views, list of relationships that should be reloaded
   * @example
   * .success('Success', { sidePanelHtml: '<blinkee>Success!</blinkee>' });
   */
  success(
    toasterMessage?: string,
    options?: { sidePanelHtml?: string; invalidated?: string[] },
  ): ActionResult {
    return {
      type: 'Success',
      message: toasterMessage ?? 'Success',
      invalidated: new Set(options?.invalidated ?? []),
      html: options?.sidePanelHtml,
    };
  }

  /**
   * Returns an error response from the action
   * @param toasterMessage the error message to return
   * @param options.sidePanelHtml in addition to the toaster, you may open a side-panel to display
   *   the result in HTML
   * @example
   * .error('Failed to refund the customer!', { sidePanelHtml: '<strong>Error!</strong>' });
   */
  error(toasterMessage?: string, options?: { sidePanelHtml: string }): ActionResult {
    return {
      type: 'Error',
      message: toasterMessage ?? 'Error',
      html: options?.sidePanelHtml,
    };
  }
```

File: docs/agent-customization/actions/responses.md [hunk 4]
```diff
@@ -44,6 +44,21 @@ return resultBuilder.success(
...
+```javascript
+const record = await context.getRecord();
+return resultBuilder.error(
+  'You can perform this action for the following reasons:',
+  `
+    <p class="c-clr-1-4 l-mt l-mb">\$${record.amount / 100} USD exceeds threshold.</p>
+    <strong class="c-form__label--read c-clr-1-2">You can use the following card instead</strong>
+    <p class="c-clr-1-4 l-mb">**** **** **** ${record.source.last4}</p>
+  `,
+  { type: 'html' },
```
Comment (romain-gilliotte) [inline docs/agent-customization/actions/responses.md:58]: doc is not up-to-date with changes made on other files

**Review by romain-gilliotte — CHANGES_REQUESTED** [review]

File: docs/agent-customization/actions/responses.md (hunk 2 above)
Comment (SteveBunlon) [reply docs/agent-customization/actions/responses.md:61]: Screenshot added 👌

File: docs/agent-customization/actions/responses.md (hunk 4 above)
Comment (SteveBunlon) [reply docs/agent-customization/actions/responses.md:58]: updated 🆗

**Review by romain-gilliotte — APPROVED** [review]

**Pull request merged**